Design for additional user wide configuration (#9658)
NuGet/Home: designs/UserWideConfiguration-AdditionalExtensibilityFolders.md

> # Spec Name

Reviewer: update? :)

Author: rip, I tried :D

1. Require SDK installers to modify the one %APPDATA%\NuGet\NuGet.config file. However this seems error prone.

## Who are the customers

All NuGet customers are affected. Primary consumers are the WPF and WinForms customers.

## Goals

* Provide a way for 3rd party vendors to add new user-wide configuration files without elevation.

## Non-Goals

## Solution

The solution to start detecting additional configs from the user specific directory similar to the extensibility from machine wide configs.

Given that these are additional configs, they will be merged with lower priority than the default user specific config. The rest are merged deterministically in alphabetical order because they are not expected to confict.
The determinism allows an out in case something really goes wrong.

There 2 solutions that satisfy the requirement with the least amount of friction.

### Re-use the existing %APPDATA%\NuGet directory

Pros:

> * Simple & intuitive. Windows: %appdata%\NuGet\NuGet.Config
> Mac/Linux: ~/.config/NuGet/NuGet.Config or ~/.nuget/NuGet/NuGet.Config.

Reviewer: Did you mean?

```suggestion
* Simple & intuitive. Windows: %appdata%\NuGet\*.Config
Mac/Linux: ~/.config/NuGet/*.[C|c]onfig or ~/.nuget/NuGet/*.[C|c]onfig.
```

Author: Yeah, that's better :D

Cons:

* This is an existing folder. There are some known instances of unexpected NuGet.config files there. Examples include: NuGet_backup.config found by the OP of the linked issue. It's also likely there are some additional copies of NuGet.config in there.

### Define a new folder for additiona user-wide configuration

> The proposal is to use `%APPDATA%\NuGet\config`.

Reviewer: ```suggestion
The proposal is to use `%APPDATA%\NuGet\config\` on Windows and `~/.config/NuGet/config/` or `~/.nuget/NuGet/config/` on Mac/Linux

```

Pros:

* Satisfies all the requirements.

Cons:

* Arguably not as intuitive as the first proposal, but it doesn't suffer from the same potential issues with unexpected configuration files. Given the difficulty one could have identifying which config brings in what source or fallback folder. The trade off might be good enough.

Worth to mention that while there have been instances of additional configs in the current user wide directory, it's impossible for us to determine the exact indicidence. We are likely talking about very few customers.

## Future Work

None

## Open Questions

> * Which of the 2 approaches do we prefer? Personally I was leaning toward re-using the same folder, but if we want to be risk averse, we can just take the new folder approach.

Reviewer: Best case: customer doesn't have any other `*.config` file in the directory, so there's no problem
Middle case: customer has a backup config file which previously was not loaded, starts getting loaded and causes restore issues. For example, maybe a backup config has a package source that no longer exists, or the customer is switching config files when on and off their corporate network that has a private package source. Restore breaks until they rename/delete the backup file(s), but it won't be at all obvious what the cause is or what the fix is, all the customer did was update VS or install a newer .NET SDK.
Worst case: there's a config file with an XML syntax error. Not only does restore break, but VS might hang too: https://github.com/NuGet/Home/issues/9574

I also prefer re-using the same folder, but I'm concerned that the "harm" for the customers with additional files on their computers might be serious enough to warrant using a subdirectory.

Author: If you had to pick, I guess you are leaning option 2 correct? I'l consider a thumbs up as yes :)

# Additional user wide configuration options

* Status: In Review
* Author(s): [Nikolche Kolev](https://github.com/nkolev92)
* Issue: [9394](https://github.com/NuGet/Home/issues/9394) NuGet should support multiple config files in %APPDATA%\NuGet directory

## Problem Background

The XAML and WinForms teams want to allow control library installers to add user-specific NuGet fallback folders by having their SDK installers drop config files in %APPDATA%\NuGet. On uninstall this file would be deleted.

We considered two alternatives which I mentioned as workarounds above:

1. Require SDK installers to drop config files in %ProgramFiles(x86)%\NuGet\Config instead. However, we've gotten feedback from control vendors that they would like the option of installing without elevation which prevents the installer from being able to write to this directory.

> 1. Require SDK installers to modify the one %APPDATA%\NuGet\NuGet.config file. However this seems error prone.

Reviewer 2: Why is this error prone? Adding one line and then removing it upon uninstall should be simple IMO. Similar to adding and removing sources/feeds.

Reviewer: My understanding is they're going to be adding a fallback folder, not package source. So if the user's config file doesn't already contain a fallback folders element, then it's 3 XML elements that needs to be added, and then there's the question whether the fallbackFolders element should be deleted if the last fallback folder is removed from that list.

If the fallbackFolders list contains multiple entries and the uninstaller has a bug, it might delete the wrong line in the file, or might delete all the fallback folders, not just its own. There's less risk that the uninstaller deletes the wrong file, or overwrites something it shouldn't on install.

Depending on what APIs the control vendor uses in their setup program, it might strip out XML comments from the file, or change whitespace. While this won't cause problems, it might not be what the customer wants, particularly for a file that is not version controlled.

Installer technologies are optimised for writing files and uninstall to delete files, so that's the easiest thing to do. Adding logic to edit existing files, while possible, increases the complexity for control vendors to use this feature.

It would be better for performance if we didn't load `*.config` from anywhere, but the risks and additional complexity for anyone who wants to use this feature doesn't feel worthwhile to me. We can minimise perf impact by loading in parallel if we're not already.

Reviewer 2: How does a Fallback folder line look in the nuget.config? If we can provide an API to add and remove the fallback folder, given a unique name, should it still be a problem. My thinking is to abstain from adding more moving parts to the nuget config management.

Author: ```xml
<fallbackPackageFolders>
   <add key="XYZ Offline Packages" value="C:\somePath\someFolder\"/>
</fallbackPackageFolders>
```

from https://docs.microsoft.com/en-us/nuget/reference/nuget-config-file#fallbackpackagefolders-section. 

Given that the installation of these configs is a 3rd party experience, they are free to use our libraries. We have them, but we don't know if the installer is written in .NET. 

Regarding the error prone part, editing onfigs has caused unexpected problems for customers in the past. 
There are some occasions in which NuGet itself trying to edit the config caused issues. 

@rrelyea might know more about the history of those issues.

Reviewer: It looks almost identical to the packageSources section: https://docs.microsoft.com/en-us/nuget/reference/nuget-config-file#fallbackpackagefolders-section.

Even if we create an API (presumably in .NET, although we technically already have APIs, in NuGet.Configuration.dll), I have no idea how installers would call the API. Can MSI's natively execute .NET assemblies, or would the MSI author have to jump through additional hoops to copy .NET assemblies to a temporary location, run, then clean up this temporary stuff after itself? What about installers that are not MSIs?

> 1. Require SDK installers to modify the one %APPDATA%\NuGet\NuGet.config file. However this seems error prone.

Reviewer 3: Another possible destination would be `%ProgramData%` folder (typically, `C:\ProgramData`). Non-admin users can write to it, and all users can read from it.

Reviewer 3: What I do not know what would be Linux/OSX equivalent? Something under `/usr`?

Author: That's certainly an option. Familiarity was the key motivator for proposing AppData. 

The programdata equivalent on linux is `/etc/opt`

Reviewer: /etc/opt on my Ubuntu on WSL is not user writable.

## Who are the customers

All NuGet customers are affected. Primary consumers are the WPF and WinForms customers.

## Goals

* Provide a way for 3rd party vendors to add new user-wide configuration files without elevation.

## Non-Goals

## Solution

> The solution to start detecting additional configs from the user specific directory similar to the extensibility from machine wide configs.

Reviewer 4: Suppose you have a project with the following layout:

```
app.config
code.cs
nuget.config
project.csproj
solution.sln
```

Would `nuget.config` and `app.config` be parsed as NuGet.config files? If that's the case, could we adopt `.nuget.config` as a file extension to avoid parsing `app.config`?

Reviewer: The proposal is to load `%appdata%\NuGet\config\*.config`, not to load `*.config` in the project directory.

Author: What @zivkan said.

Reviewer 2: What is the expected behavior if </clear> is mentioned in multiple files? I feel irrespective of the priority defined by us, its going to be confusing.

> Given that these are additional configs, they will be merged with lower priority than the default user specific config. The rest are merged deterministically in alphabetical order because they are not expected to confict.

Reviewer 4: ```suggestion
Given that these are additional configs, they will be merged with lower priority than the default user specific config. The rest are merged deterministically in lexicographical order because they are not expected to conflict.
```

Reviewer 4: - [ ] Reminder for Fernando: Make sure that ordering is culture invariant.

The determinism allows an out in case something really goes wrong.

> There 2 solutions that satisfy the requirement with the least amount of friction.

Reviewer 4: ```suggestion
There are 2 solutions that satisfy the requirement with the least amount of friction.
```

### Re-use the existing %APPDATA%\NuGet directory

Pros:

* Simple & intuitive. Windows: %appdata%\NuGet\*.Config
Mac/Linux: ~/.config/NuGet/*.[C|c]onfig or ~/.nuget/NuGet/*.[C|c]onfig.

Cons:

* This is an existing folder. There are some known instances of unexpected NuGet.config files there. Examples include: NuGet_backup.config found by the OP of the linked issue. It's also likely there are some additional copies of NuGet.config in there.

> ### Define a new folder for additiona user-wide configuration

Reviewer 4: ```suggestion
### Define a new folder for additional user-wide configuration
```

The proposal is to use `%APPDATA%\NuGet\config\` on Windows and `~/.config/NuGet/config/` or `~/.nuget/NuGet/config/` on Mac/Linux

Pros:

* Satisfies all the requirements.

Cons:

* Arguably not as intuitive as the first proposal, but it doesn't suffer from the same potential issues with unexpected configuration files. Given the difficulty one could have identifying which config brings in what source or fallback folder. The trade off might be good enough.

Worth to mention that while there have been instances of additional configs in the current user wide directory, it's impossible for us to determine the exact indicidence. We are likely talking about very few customers.

## Future Work

> None

Reviewer 4: Would we detect `<clear/>` nodes in the installer's config file?

Author: The config merges works the same way it does today for any number of configs. 

This only affects the config discovery, so all `</clear>` tags will work as before :)
